### PR TITLE
RSDK-3036 - Make move functions stoppable

### PR DESCRIPTION
**Pull request by kkufieta**
Ticket: https://viam.atlassian.net/browse/RSDK-3036

Done:
* Added `inputEnabledActuator` interface to the builtin motion service
* Made the following functions stoppable: `Move`, `MoveSingleComponent`, `MoveOnMap`, `MoveOnGlobe`
* Extended injected Arm with `CurrentInputsFunc` and `GoToInputsFunc`
* Wrote tests & refactored existing tests so I could reuse code

### CODE REVIEW

File: services/motion/builtin/builtin.go [hunk 1]
```diff
@@ -157,20 +159,27 @@ func (ms *builtIn) Move(
...
-	for _, step := range output {
+	for _, step := range steps {
 		// TODO(erh): what order? parallel?
 		for name, inputs := range step {
 			if len(inputs) == 0 {
 				continue
 			}
 			err := resources[name].GoToInputs(ctx, inputs)
 			if err != nil {
+				for _, step := range steps {
```
Comment (raybjork) [inline services/motion/builtin/builtin.go:176]: I think theres probably better way to write this than nesting this far in, but the idea is right

**Review by raybjork — COMMENTED** [review]
This looks ok but I'm not sure its complete as written.  To validate we will want unit tests to prove this works as expected.  For example using inject to set up a situation where we are trying to move a few arms (and instead of them actually moving they just waste cycles), then cancelling the context of the Move should get all the components to return rather than orphaning any processes

File: components/arm/arm.go [hunk 2]
```diff
@@ -65,8 +65,7 @@ type Arm interface {
 	resource.Resource
 	referenceframe.ModelFramer
 	resource.Shaped
-	resource.Actuator
```
Comment (randhid) [inline components/arm/arm.go:68]: I'm sorry - I'd rather not have the actuator interface be nested in Input enabled in some resources (arm, gantry, kinematic base) and not in others (motor, servo, base). Could we keep it flat?

Comment (kkufieta) [reply components/arm/arm.go:68]: @raybjork Do you want to weigh in on this?

Comment (raybjork) [reply components/arm/arm.go:68]: Is this objection from a readability perspective Rand?

Comment (raybjork) [reply components/arm/arm.go:68]: @randhid and I talked offline and agreed on a compromise solution, renaming this interface to `InputEnabledActuator` I want to tag @biotinker on this too since he might have thoughts

**Review by randhid — APPROVED** [review]
Looks good! Have you tested it out on something with a move call?

File: components/arm/arm.go (hunk 2 above)
Comment (biotinker) [reply components/arm/arm.go:68]: The addition of `resource.InputEnabledActuator` as a wrapper seems weird to me.

I think it would be cleaner to move `referenceframe.InputEnabled` to `resource`, and then simply keep it separate from `Actuator`.

Comment (raybjork) [reply components/arm/arm.go:68]: This is coming up because we want to be able to more easily reference things we move in the motion service. Right now we store them as input enabled and have to assert they are actuators, but if we had one type that fully encompasses things we can move that would be cleaner

Comment (biotinker) [reply components/arm/arm.go:68]: That makes sense.
Another option is to put this interface privately on the motion service itself. It can cast everything to its own internal type then, once.

Also OK with this as-is but it seems somewhat clunky.

Comment (randhid) [reply components/arm/arm.go:68]: that actually seems like the best option, thanks for the suggestion peter!

File: testutils/inject/base.go [hunk 3]
```diff
@@ -33,7 +34,10 @@ func NewBase(name string) *Base {
 
 // Name returns the name of the resource.
 func (b *Base) Name() resource.Name {
-	return b.name
+	if b.NameFunc == nil {
```
Comment (randhid) [inline testutils/inject/base.go:37]: why is this needed, and why can't you use the resource name of the injected base?

Comment (kkufieta) [reply testutils/inject/base.go:37]: I am still working on this. It's not yet ready for review yet. I added this because I couldn't get my tests working without it. Now that my tests work, I'll look into this more closely.

Comment (kkufieta) [reply testutils/inject/base.go]: Done now :) Got rid of this again.

**Comment by kkufieta** [comment]
> Have you tested it out on something with a move call?

@randhid I added tests but didn't have the chance to test this with hardware. Are the unit tests sufficient?

**Comment by raybjork** [comment]
> > Have you tested it out on something with a move call?
> 
> @randhid I added tests but didn't have the chance to test this with hardware. Are the unit tests sufficient?

I can do some tests on hardware, thanks for finishing this ticket up Kat!

File: services/motion/builtin/builtin_test.go [hunk 4]
```diff
@@ -652,3 +677,200 @@ func TestGetPose(t *testing.T) {
 	test.That(t, err, test.ShouldBeError, referenceframe.NewParentFrameMissingError("testFrame", "noParent"))
 	test.That(t, pose, test.ShouldBeNil)
 }
+
+func TestStoppableMoveFunctions(t *testing.T) {
+	ctx := context.Background()
+	logger := golog.NewTestLogger(t)
+	failToReachGoalError := errors.New("failed to reach goal")
+	calledStopFunc := false
```
Comment (raybjork) [inline services/motion/builtin/builtin_test.go:685]: I think this line isn't necessary

**Review by raybjork — APPROVED** [review]

File: services/motion/builtin/builtin_test.go (hunk 4 above)
Comment (kkufieta) [reply services/motion/builtin/builtin_test.go:685]: Do you mean because `calledStopFunc` could be (and partially is already) defined within each `t.Run`?

**Review by biotinker — APPROVED** [review]
LGTM, request for additional comments but non-blocking.

File: services/motion/builtin/builtin_test.go (hunk 4 above)
Comment (raybjork) [reply services/motion/builtin/builtin_test.go:685]: Yeah that was my reading, but either way its fine

Comment (kkufieta) [reply services/motion/builtin/builtin_test.go:685]: Turns out it's necessary for the function definition right after.

**Review by randhid — APPROVED** [review]
